[QoI] Improve diagnostics for type member names shadowing top-level functions

xedin opened the pull request:
Consider expression with an implicit 'self.' reference:
```swift
extension Sequence {
  func test() -> Int {
    return max(1, 2)
  }
}
```
We currently shadow names that appear in nested scopes, so the
top-level two-argument min()/max() functions are shadowed by the
versions of min()/max() in Sequence (which don’t take the same
number of arguments). This patch aims to improve situation on this
front and produce better diagnostics when that happens, hinting
the user about what went wrong and where matching function is
declared.

Resolves <rdar://problem/25341015>.
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
/cc @rudkx

jtbandes commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -825,6 +825,9 @@ ERROR(argument_out_of_order_unnamed_unnamed,none,
 
+ERROR(func_is_not_an_instance_member,none,
+      "%0 with given arguments is not an instance member and "
+      "cannot be called using implicit 'self.'", (DeclName))
```
IMO, since the implicit `self.` is implicit, it shouldn't be mentioned to the user (since they didn't write it). Something like `static member 'MyClass.max' shadows global function 'Swift.max'` might be more intuitive, especially if you could provide fix-it hints for inserting the `MyClass.` and/or `Swift.` (or whichever module the other candidate came from).

jtbandes commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4555,6 +4555,80 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+                                     /*isSpecialized=*/false,
+                                     UDE->getFunctionRefKind()));
+
+  CalleeCandidateInfo candidates(nullptr, choices,hasTrailingClosure, CS,
```
nit: missing space in `choices,hasTrailingClosure`

jtbandes commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4555,6 +4555,80 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+  //
+  // In this example Sequence class already has two methods named 'max'
+  // none of which accept two arguments, but there is a function in
+  // Swift Standard Library called 'max' does which accepts two arguments,
```
nit: `does which accepts` → `which does accept` or `which accepts`

jtbandes commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4555,6 +4555,80 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+  //   }
+  // }
+  //
+  // In this example Sequence class already has two methods named 'max'
```
nit: `Sequence class` → `the Sequence class`

jtbandes commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4555,6 +4555,80 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+  auto moduleCtx = baseDecl->getDeclContext()->getModuleScopeContext();
+  auto result = TC.lookupUnqualified(moduleCtx, UDE->getName(), UDE->getLoc());
+
+  // Let's type check expression using original list of the candidates.
```
nit: `expression` → `the expression`

jtbandes commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4555,6 +4555,80 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+    return false;
+
+  // Now let's construct a list of the globally scoped candidates,
+  // match argument expression against that and check if it does match.
```
nit: `argument` → `the argument`

xedin commented:
@jtbandes Thanks for the review! I've corrected the spelling and changed message to look like this:

```
error: member 'Sequence.max' shadows global function 'Swift.max'
       return max(1, 2)
              ^
              Swift.
Swift.max:7:13: note: 'max' declared here
public func max<T>(_ x: T, _ y: T) -> T where T : Comparable
```

rintaro commented:
This PR crashes in this case:
```swift
class Foo {
  func foo() {

    func baz(_ x: Int, _ y: Int) {}

    class Bar {
      func baz() {}
      func qux() {
        baz(1, 2)
      }
    }

  }
}
```

jtbandes commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4555,6 +4555,105 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+              getName(baseCtx), moduleName, UDE->getName().getBaseName().str())
+      .fixItInsert(baseExpr->getEndLoc(), moduleName + ".");
+
+  for (auto &candidate : candidates.candidates) {
```
nit: `const auto &`

jtbandes commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -825,6 +825,10 @@ ERROR(argument_out_of_order_unnamed_unnamed,none,
       (unsigned, unsigned))
 
+ERROR(member_shadows_global_function,none,
+      "member '%0.%2' shadows global function '%1.%2'",
```
Thanks, I think this will be much more helpful to users. Thinking about it more, it's still not quite enough information to allow someone to understand *why* it's an error. (If %0 shadows %1, why doesn't the compiler just treat it as a call to %0?)

An idea to make this even better: change `member` → `static member` (or perhaps `class member` as appropriate, or just `type member`), and add `, but cannot be called with implicit 'self.'` similar to what you had originally.

It might also be nice to include a fix-it that inserts `MyType.` or `type(of: self).` (assuming those would work).

jrose-apple replied:
We also have DescriptiveDeclKind, for standardized user-facing descriptions of decls. That'll give you something like "class method".

jrose-apple replied:
(It's not special-cased for "property" and "class property" yet, though. We should probably do that.)

xedin replied:
@jrose-apple thanks for the pointer, using DescriptiveDeclKind would make the message much better!

jtbandes commented on test/Constraints/members.swift:
```diff
@@ -590,3 +590,10 @@ do {
+// rdar://problem/25341015
+extension Sequence {
+  func r25341015() -> Int {
+    return max(1, 2) // expected-error {{member 'Sequence.max' shadows global function 'Swift.max'}}
```
With a custom protocol you could also demonstrate the `decl_declared_here` note. It might be best for the test to include an example with more than just 2 candidates.

jtbandes commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4555,6 +4555,105 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+  // Get outermost context relative to this expression.
+  auto baseCtx = baseDecl->getDeclContext();
+  auto moduleCtx = baseCtx->getModuleScopeContext();
+  auto result = TC.lookupUnqualified(moduleCtx, UDE->getName(), UDE->getLoc());
```
Maybe a silly question, but are we guaranteed that the results of this are exactly the same as the regular lookup that occurs (outside of attempting to diagnose this error)? I'm just wondering whether it's possible that this fix-it code would produce a qualified call that doesn't actually reference an accessible function.

xedin commented:
@rintaro Thanks for reporting it I'll fix it asap!
@jtbandes Thanks for the review once again! I'll extend the message and add couple more tests (similar to what @rintaro posted). Regarding your question - since lookup is going to be done on the module level it would definitely find more results then original contextual one produced.

xedin commented:
@jtbandes I've updated the algorithm to walk up the list of the contexts and try to diagnose shadowing on each one (only for static and global functions) and I've also updated the message to what you have suggested and added fix-it and couple of new tests. Please take a look!

Here is an example:
```
(swift) class Foo {
            static func baz(_ x: Int, _ y: Int) {}
            func baz() {}
            func qux() {
              baz(1, 2)
          }
        }
<REPL Input>:5:7: error: class member 'baz' shadows static method 'baz' in class 'Foo', which cannot be called using implicit 'self.'
      baz(1, 2)
      ^
      Foo.
<REPL Input>:2:17: note: 'baz' declared here
    static func baz(_ x: Int, _ y: Int) {}
                ^
```

xedin commented:
@rintaro I've added both of your examples as tests, thank you once again!

rudkx commented:
@xedin Can you re-title to "QoI" rather than "QoL" and I'll try to take a look today.

xedin commented:
@rudkx Done and thanks! I've pushed everything so all involved can take a look at the updated diagnostic, haven't run all validation tests, so something might be broken...

rudkx commented:
@swift-ci Please smoke test

xedin commented:
@rudkx Looks like I broke one of the tests, will investigate tonight and let you know.

xedin commented:
@rudkx I've fixed the issue which re-introduced crasher (base had to be validated before considering a candidate), -T passes for me now.

jtbandes commented:
Commit message still has a lowercase `L`, should be a capital `i` 😄

xedin commented:
@jtbandes Ah, I thought it stands for "Quality of Language" that's why I kept on putting "l" everywhere... Fixed now, thanks!

modocache commented:
"QoI" confused me as well. It's [in the Swift lexicon](https://github.com/apple/swift/blame/93aa6bfef5487384537ff1b8d1fe767792512910/docs/Lexicon.rst#L191-L199), for what it's worth, but that's probably a bit hard to find...

jtbandes commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -825,6 +825,11 @@ ERROR(argument_out_of_order_unnamed_unnamed,none,
       (unsigned, unsigned))
 
+ERROR(member_shadows_global_function,none,
+      "class member %0 shadows %1 %0 in %2 '%3'"
```
You've hard-coded `class member`, but couldn't the same problem happen for a struct or enum, or perhaps even a protocol or protocol extension?

xedin commented:
@modocache Thanks, that's really helpful!

xedin commented:
@jtbandes You are right, it might happen with anything nominal, I've changed message to say "type member" because it's hard to get the base type and added test for struct.

jtbandes commented on test/Constraints/members.swift:
```diff
@@ -590,3 +590,36 @@ do {
+// rdar://problem/25341015
+extension Sequence {
+  func r25341015_1() -> Int {
+    return max(1, 2) // expected-error {{type member 'max' shadows global function 'max' in module 'Swift', which cannot be called using implicit 'self.'}} {{12-12=Swift.}}
```
Sorry to keep bikeshedding this so much 😝

The message currently implies that `Swift.max` cannot be called using implicit `self.`. But that's not what the code is trying to do.

One way of solving this linguistic problem would be replacing `which` with either `and` or `but`, so it's clear that the type member is the one that can't be called with implicit `self.`.

However, it might be clearer overall if you rearrange the clauses:

> `global function 'Swift.max' is shadowed by instance member 'max', which cannot be called using implicit 'self.'`

What do you think? (Is it correct to say `instance member`? I think that's clearer than `type member`, especially when the instance member is conflicting with a static member of the same type).

jtbandes replied:
Actually, I think I was wrong about this.

When a global function conflicts with an instance member, `which cannot be called using implicit 'self.'` is a bit confusing.

But when a *static member* conflicts with an instance member, the message is correct. So, maybe you can just disregard what I said above...

xedin replied:
@jtbandes Let's see what @rudkx and others have to say about the message and I will change it accordingly :)

xedin replied:
After thinking about this some more it might be better to switch from `type member` to `instance member` since that's going to be as accurate as it gets (because the declaration of the member can't be easily resolved)...

jtbandes replied:
IIRC from thinking about this before, the problem was that the conflicting member isn't *always* an instance member. It could be an instance member conflicting with a global function, or a static member conflicting with a global function, right?

xedin replied:
You are right, for some reason I thought that "." is involved in here... The problem is that member declaration is unresolved which means we can't really say if it's (supposed to be) a static or instance member at this time, so it's going to be hard to differentiate messages...

jrose-apple replied:
I think mentioning "implicit self" doesn't help at all—it's clearly not what the user was intending to do if we get into this situation. Maybe something more like "use of 'max' refers to instance property 'max' rather than"

Actually, we were in this situation before with NSView.print(\_:) (5c71b75b250ef5d564a9bc7bc451fb72227de2fd). I think I like the message format we went with there:

```
<stdin>:5:5: warning: use of 'print' treated as a reference to instance method in class 'NSView'
    print("hi")
    ^
AppKit.NSView:7:15: note: 'print' declared here
    open func print(_ sender: Any?)
              ^
<stdin>:5:5: note: use 'self.' to silence this warning
    print("hi")
    ^
    self.
<stdin>:5:5: note: use 'Swift.' to reference the global function
    print("hi")
    ^
    Swift.
```

rintaro commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4555,6 +4555,188 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+      return getBaseName(context->getParent());
+
+    case swift::DeclContextKind::AbstractFunctionDecl:
+      return cast<AbstractFunctionDecl>(context)->getNameStr();
```
This path is reachable? If so, test case for this?

If not, this `getBaseName` can be implemented as:
```cpp
if (context->isTypeContext())
  return context->getAsGenericTypeOrGenericTypeExtensionContext()->getNameStr();
else if (context->isModuleScopeContext())
  return context->getParentModule()->getNameStr();
else
  llvm_unreachable("...")
```

xedin replied:
I don't think it's reachable, I think it can only be if we were to accept local functions. Thanks for the suggestion, @rintaro, I'll change modify the code accordingly.

xedin replied:
@rintaro I've made changes to the getBaseName and rebased everything with the latest master.

jrose-apple replied:
We do accept local functions.

```swift
func test() {
  func inner() { print("hi") }
  inner()
}
```

rudkx commented:
@swift-ci Please smoke test

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -825,6 +825,11 @@ ERROR(argument_out_of_order_unnamed_unnamed,none,
       (unsigned, unsigned))
 
+ERROR(member_shadows_global_function,none,
+      "type member %0 shadows %1 %0 in %2 '%3'"
```
Additionally, the full name of the member may be different than the full name of the thing being shadowed, so reusing `%0` is incorrect.

xedin commented:
@jrose-apple Thanks for looking at this! I've addressed all of your points and I like the message format you suggested so I made it work too, here is what I've got (not sure is `CCI[0]->getDescriptiveDeclKind()` is ok or not through):

```
(swift) extension Sequence {
            func r25341015_1() -> Int {
              return max(1, 2)
          }
        }
<REPL Input>:3:14: error: use of 'max' refers to instance method 'max' rather than global function in module 'Swift'
      return max(1, 2)
             ^
<REPL Input>:3:14: note: use 'Swift.' to reference the global function in module 'Swift'
      return max(1, 2)
             ^
             Swift.
Swift.max:7:13: note: 'max' declared here
public func max<T>(_ x: T, _ y: T) -> T where T : Comparable
            ^
```

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -836,6 +836,11 @@ ERROR(argument_out_of_order_unnamed_unnamed,none,
       (unsigned, unsigned))
 
+ERROR(member_shadows_global_function,none,
+      "use of %1 refers to %0 %1 rather than %2 in %3 %4",
```
I think this should use the full name for the member, so "use of %0 refers to %1 %2 rather than %3 in %4 %5".

xedin replied:
Will do, thanks!

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4584,6 +4584,178 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+    auto baseName = getBaseName(choice->getDeclContext());
+
+    TC.diagnose(UDE->getLoc(), diag::member_shadows_global_function,
+                CCI[0].getDecl()->getDescriptiveKind(), choice->getName(),
```
`choice->getDescriptiveDeclKind()`? and `choice->getDeclName()`, once you've done the diagnostic change I mentioned.

xedin replied:
`CCI` is a "original" candidates (on the nominal) and `candidates` and `choice` are lookup on the parent contexts, so this message needs both.

xedin replied:
and `choice` doesn't have `getDeclName` did you mean `getBaseName` (which is the same as `getName`)?

jrose-apple replied:
Ah, `getFullName`, sorry. The one that actually gives you the name with the argument labels included.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -4584,6 +4584,178 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+                                   CCI.hasTrailingClosure, CS, base);
+
+    calleeInfo.filterList(argType, argLabels);
+    if (calleeInfo.closeness != CC_ExactMatch)
```
This guarantees that there's at least one, right?

xedin replied:
I'll add additional check for CCI being empty at the top, because calleeInfo and CCI are to different things actually.

jrose-apple commented:
(I do like the new wording, and just picking the first candidate seems acceptable in this case.)

xedin commented:
@jrose-apple 

I've corrected message to use labels and reformatted it as you suggested, plus added
labels for the `struct` test case. Regarding local function test case I'm not sure what exactly
is required but there is one already included (the very last one)...

I think I made it right this time, here is the output: 

```
struct S_25341015 {
            static func foo(_ x: Int, y: Int) {}

            func foo() {}
            func bar() {
              foo(1, y: 2)
          }
        }
<REPL Input>:6:7: error: use of 'foo' refers to instance method rather than static method 'foo(_:y:)' in struct 'S_25341015'
      foo(1, y: 2)
      ^
<REPL Input>:6:7: note: use 'S_25341015.' to reference the static method
      foo(1, y: 2)
      ^
      S_25341015.
<REPL Input>:2:17: note: 'foo(_:y:)' declared here
    static func foo(_ x: Int, y: Int) {}
                ^
```

jrose-apple commented:
This is the test I wanted:

```swift
func foo(x: Int, y: Int) {}
func outer() {
  func foo() {}
  foo(x: 1, y: 2)
}
```

It's okay that your logic doesn't produce a better message here, but it's still worth testing.

jrose-apple commented:
```
<REPL Input>:6:7: error: use of 'foo' refers to instance method rather than static method 'foo(_:y:)' in struct 'S_25341015'
```

Heh, I actually wanted the full name after the thing it did find (in this case, the "instance method"). I don't care as much whether it's on the other one because presumably that's the one you wanted to begin with, but sure, why not.

xedin commented:
```
Heh, I actually wanted the full name after the thing it did find (in this case, the "instance method")
```

I think that's where the disconnect is because what it did find is actually "static method foo(_:y:)", `instance method` (identified by CCI[0] passed in) is that it _tried_ to find originally and failed because of the incorrect arguments or something else...

jrose-apple commented:
Consider this case:

```swift
func foo(x: Int, y: Int) {}
struct X {
  func foo(z: Int) {}
  func test() {
    foo(x: 1, y: 2)
  }
}
```

I want the error message to be "use of 'foo' refers to instance method 'foo(z:)' rather than global function in module 'main'" (or possibly "global function 'foo(x:y:)' in module 'main'). I think it's important to say what the compiler found specifically, since it's not what the user intended.

xedin commented:
I understand your point completely, but I just don't know what it should output if situation is like this e.g.

```swift

```

Since there are multiple "original" candidates with different labels do we simply omit name of the declaration?

jrose-apple commented:
Ah. If there are multiple candidates, even saying "instance method" might be a lie (instance methods and properties can be overloaded in some cases). So I think just picking the first one seems fine.

xedin commented:
Ok, thanks for clarifying, @jrose-apple! I'll always pick the first one from CCI and use it no matter how many candidates there are.

xedin commented:
@jrose-apple Done! Changed the message to include the full name of the member and added missing test case for the local function.

jrose-apple commented:
@swift-ci Please test

xedin commented:
@jrose-apple Please re-run, this happened because of all force pushing :(

jrose-apple commented:
It automatically reruns whenever there's a wrong SHA failure. I'll ask Mishal if we can get it to not comment in that case…

xedin commented:
Good to know!

xedin commented:
Ok, looks like CI failed and I can reproduce after rebasing with the latest master, will dig into that soon.

xedin commented:
@jrose-apple I've moved to using correct `getInterfaceType` instead of `getType` on type decls, so local -T is all green now, can you please re-run CI?

rudkx commented:
@swift-ci Please smoke test

xedin commented:
@rudkx Thank you!
